날짜 | 범위
--- | ---
1/26 | 11.3 프라미스 체이닝 ~ 11.5 프라미스 API

### 프라미스 체이닝
- 프라미스 체이닝은 `result`가 `.then` 핸들러의 체인을 통해 전달된다.
- 프라미스 객체를 생성해 `then`을 여러개 추가하는 것은 체이닝이 아니다. 그렇게 되면 여러개의 `then`이 하나의 `result`를 가지고 동시에 실행되는 것이다.
- `then` 핸들러 안의 `return`문에서 프라미스를 새로 생성하여 리턴하는 체이닝을 구성할 수도 있다.
- 핸들러는 프라미스 객체가 아닌 `thenable` 객체를 반환하기도 한다. `thenable`은 `then` 메서드가 구현되어 있는 객체를 일컫는다.
```javascript
let promise = fetch(url);
//네트워크 요청으로 fetch를 활용하여 프라미스를 활용

fetch(url)
  .then(response => response.json())  //response.json()로 파싱
  .then(json => new Promise((res, rej) => {
    setTimeout(() => {
      res(json.name);
    }, 3000);
  })
  .then(result => console.log(result));
```
- `then` 핸들러 내부에서 `setTimeout`을 실행하고 실행이 완료된 뒤 무언가를 추가하고 싶다면 프라미스를 반환해서 그다음 `then`으로 실행하면 된다.

### 프라미스와 에러 핸들링
- `.catch`는 프라미스가 거부되면(`reject`) 호출되고, 여러개의 `then` 뒤에 올 수 있다. (여러개의 `then`에서 발생할 수 있는 모든 에러를 하나의 `catch`가 잡는다)
- `catch`에서 에러가 정상적으로 처리가 되면 그 다음 가까운 `then` 핸들러로 넘어간다.
- `catch`에서 처리하지 못한 에러를 `throw error`로 넘긴다면 그 다음 `catch`에서 에러를 확인한다.
- 브라우저에서는 `catch`에서 처리하지 못한 에러를 `unhandledrejection` 이벤트 핸들러로 추적할 수 있다.
- `catch`는 동기적 에러만 처리할 수 있다.(`setTimeout`과 같이 비동기 코드를 실행하다 에러가 난 경우 `catch`가 실행되지 않는다)

### 프라미스 API
```javascript
Promise.all([...promise...]);
/* 
- all은 여러개의 프라미스 배열을 요청하고 모두 완료된 후 결과값을 배열로 받는다. 배열의 순서는 요청된 프라미스 순서와 같다.
- 프라미스 중 하나라도 에러가 나면 전체가 거부되어 catch가 실행된다. 프라미스 실행이 취소되진 않지만 결과가 무시된다.
*/

Promise.allSettled([...promise...]);
/*
- settled는 모든 프라미스가 처리될 때까지 기다렸다가 결과를 반환한다. 
- 성공 : { status: 'fulfilled', value: resule }
- 실패 : { status: 'rejected', reason: error }
*/

Promise.race(iterable);
/*
- all과 마찬가지로 여러개의 프라미스를 호출할 수 있고 제일 먼저 처리되는 프라미스의 결과(성공 혹은 에러)를 반환한다.
*/
```
- `Promise.resolve(value)`와 `Promise.reject(error)`보다 `async/await` 문법을 더 많이 쓴다.
